sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala

```diff
@@ -207,3 +247,203 @@ trait InsertableRelation {
+ *        directories of all partition directories.
+ * @param maybePartitionSpec An [[FSBasedRelation]] can be created with an optional
+ *        [[PartitionSpec]], so that partition discovery can be skipped.
+ */
+@Experimental
+abstract class FSBasedRelation private[sql](
+    val paths: Array[String],
+    maybePartitionSpec: Option[PartitionSpec])
+  extends BaseRelation {
+
+  /**
+   * Constructs an [[FSBasedRelation]].
+   *
+   * @param paths Base paths of this relation.  For partitioned relations, it should be either root
+   *        directories of all partition directories.
+   * @param partitionColumns Partition columns of this relation.
+   */
+  def this(paths: Array[String], partitionColumns: StructType) =
+    this(paths, {
+      if (partitionColumns.isEmpty) None
+      else Some(PartitionSpec(partitionColumns, Array.empty[Partition]))
+    })
+
+  /**
+   * Constructs an [[FSBasedRelation]].
+   *
+   * @param paths Base paths of this relation.  For partitioned relations, it should be either root
+   *        directories of all partition directories.
+   */
+  def this(paths: Array[String]) = this(paths, None)
+
+  private val hadoopConf = new Configuration(sqlContext.sparkContext.hadoopConfiguration)
+
+  private var _partitionSpec: PartitionSpec = _
+  refreshPartitions()
+
+  private[sql] def partitionSpec: PartitionSpec = _partitionSpec
+
+  private[sql] def refresh(): Unit = {
+    refreshPartitions()
+  }
+
+  private def refreshPartitions(): Unit = {
+    _partitionSpec = maybePartitionSpec.getOrElse {
+      val basePaths = paths.map(new Path(_))
+      val leafDirs = basePaths.flatMap { path =>
+        val fs = path.getFileSystem(hadoopConf)
+        if (fs.exists(path)) {
+          SparkHadoopUtil.get.listLeafDirStatuses(fs, fs.makeQualified(path))
+        } else {
+          Seq.empty[FileStatus]
+        }
+      }.map(_.getPath)
+
+      if (leafDirs.nonEmpty) {
+        PartitioningUtils.parsePartitions(leafDirs, "__HIVE_DEFAULT_PARTITION__")
+      } else {
+        PartitionSpec(StructType(Array.empty[StructField]), Array.empty[Partition])
+      }
+    }
+  }
+
+  /**
+   * Schema of this relation.  It consists of [[dataSchema]] and all partition columns not appeared
+   * in [[dataSchema]].
+   */
+  override lazy val schema: StructType = {
+    val dataSchemaColumnNames = dataSchema.map(_.name.toLowerCase).toSet
+    StructType(dataSchema ++ partitionSpec.partitionColumns.filterNot { column =>
+      dataSchemaColumnNames.contains(column.name.toLowerCase)
+    })
+  }
+
+  /**
+   * Specifies schema of actual data files.  For partitioned relations, if one or more partitioned
+   * columns are contained in the data files, they should also appear in `dataSchema`.
+   */
+  def dataSchema: StructType
+
+  /**
+   * For a non-partitioned relation, this method builds an `RDD[Row]` containing all rows within
+   * this relation. For partitioned relations, this method is called for each selected partition,
```

Comment:
Yes. For partitioned tables, `buildScan` is called once for each selected partition. Then all scanned rows are combined together with `unionAll` to form the final result.